src/friendly-iframe-embed.js

```diff
@@ -0,0 +1,223 @@
+ * @return {!Promise<FriendlyIframeEmbed>}
+ */
+export function installFriendlyIframeEmbed(iframe, container, spec) {
+  const win = getTopWindow(iframe.ownerDocument.defaultView);
+  const extensions = extensionsFor(win);
+
+  iframe.style.visibility = 'hidden';
+  iframe.setAttribute('referrerpolicy', 'unsafe-url');
+
+  // Pre-load extensions.
+  if (spec.extensionIds) {
+    spec.extensionIds.forEach(
+        extensionId => extensions.loadExtension(extensionId));
+  }
+
+  const html = mergeHtml(spec);
+
+  // Receive the signal when iframe is ready: it's document is formed.
+  iframe.onload = () => {
+    // Chrome does not reflect the iframe readystate.
+    iframe.readyState = 'complete';
+  };
+  let readyPromise;
+  if (isSrcdocSupported()) {
+    iframe.srcdoc = html;
+    // TODO(dvoytenko): Look for a way to get a faster call from here.
+    // Experiments show that the iframe's "load" event is consistently 50-100ms
+    // later than the contentWindow actually available.
+    readyPromise = loadPromise(iframe);
+    container.appendChild(iframe);
+  } else {
+    iframe.src = 'about:blank';
+    container.appendChild(iframe);
+    const childDoc = iframe.contentWindow.document;
+    childDoc.open();
+    childDoc.write(html);
+    childDoc.close();
+    // Window is created synchornously in this case.
+    readyPromise = Promise.resolve();
+  }
+  return readyPromise.then(() => {
+    // Add extensions.
+    extensionsFor(win).installExtensionsInChildWindow(
+        iframe.contentWindow, spec.extensionIds || []);
+    // Ready to be shown.
+    iframe.style.visibility = '';
+    return new FriendlyIframeEmbed(iframe, spec);
+  });
+}
+
+
+/**
+ * Merges base and fonts into html document.
+ * @param {!FriendlyIframeSpec} spec
+ */
+function mergeHtml(spec) {
+  const originalHtml = spec.html;
+  const originalHtmlUp = originalHtml.toUpperCase();
+
+  // Find the insertion point.
+  let ip = originalHtmlUp.indexOf('<HEAD');
+  if (ip != -1) {
+    ip = originalHtmlUp.indexOf('>', ip + 1) + 1;
+  }
+  if (ip == -1) {
+    ip = originalHtmlUp.indexOf('<BODY');
+  }
+  if (ip == -1) {
+    ip = originalHtmlUp.indexOf('<HTML');
+    if (ip != -1) {
+      ip = originalHtmlUp.indexOf('>', ip + 1) + 1;
+    }
+  }
+
+  const result = [];
+
+  // Preambule.
+  if (ip > 0) {
+    result.push(originalHtml.substring(0, ip));
+  }
+
+  // Add <BASE> tag.
+  result.push(`<base href="${escapeHtml(spec.url)}">`);
+
+  // Load fonts.
+  if (spec.fonts) {
+    spec.fonts.forEach(font => {
+      result.push(
+          `<link href="${escapeHtml(font)}" rel="stylesheet" type="text/css">`);
+    });
+  }
+
+  // Postambule.
+  if (ip > 0) {
+    result.push(originalHtml.substring(ip));
+  } else {
+    result.push(originalHtml);
+  }
+
+  return result.join('');
+}
+
+
+/**
+ * Exposes `mergeHtml` for testing purposes.
+ * @param {!FriendlyIframeSpec} spec
+ * @visibleForTesting
+ */
+export function mergeHtmlForTesting(spec) {
```

Comment:
`mergeHtmlForTesting` will be DCE'd. This way:

1. I don't need to worry about anyone ignoring `@visibleForTesting`
2. `mergeHtml` will be obfuscated.